File: Cn Python Codes/selective repeat.py
# SELECETIVE REPEAT
import random

def selective_repeat_transmission(i, N, tf):
    tt = 0  # Total frames sent and resent
    ack = [False] * (tf + 1)  # List to keep track of which frames are acknowledged

    while i <= tf:
        for k in range(i, min(i + N, tf + 1)):
            if not ack[k]:
                print(f"Sending Frame {k}")
                tt += 1  # Increase the count of total transmissions

        for k in range(i, min(i + N, tf + 1)):
            if not ack[k]:
                f = random.randint(0, 1)  # Simulate acknowledgment success/failure
                if f == 0:
                    print(f"Acknowledgment for Frame {k}")
                    ack[k] = True  # Mark frame as acknowledged
                else:
                    print(f"Timeout!! Frame Number: {k} Not Received")

        print()

        while i <= tf and ack[i]:
            i += 1  # Move the window forward for acknowledged frames

    return tt

def main():
    tf = int(input("Total number of frames: "))
    N = int(input("Window Size: "))
    i = 1
    tt = selective_repeat_transmission(i, N, tf)
    print(f"Total number of frames which were sent and resent are: {tt}")

if __name__ == "__main__":
    main()

File: selective repeat.java
# SELECETIVE REPEAT
import java.util.Random;
import java.util.Scanner;

public class Selective_repeat {

    public static int selectiveRepeatTransmission(int i, int N, int tf) {
        int tt = 0;  // Total frames sent and resent
        boolean[] ack = new boolean[tf + 1];  // Acknowledgment array
        
        Random rand = new Random();

        while (i <= tf) {
            // Sending frames in the window
            for (int k = i; k < Math.min(i + N, tf + 1); k++) {
                if (!ack[k]) {
                    System.out.println("Sending Frame " + k);
                    tt++;  // Count the frame sent
                }
            }

            // Simulating acknowledgments for the sent frames
            for (int k = i; k < Math.min(i + N, tf + 1); k++) {
                if (!ack[k]) {
                    int f = rand.nextInt(2);  // Random acknowledgment: 0 = success, 1 = failure
                    if (f == 0) {
                        System.out.println("Acknowledgment for Frame " + k);
                        ack[k] = true;  // Frame acknowledged
                    } else {
                        System.out.println("Timeout!! Frame Number: " + k + " Not Received");
                    }
                }
            }

            System.out.println();

            // Slide the window forward for acknowledged frames
            while (i <= tf && ack[i]) {
                i++;
            }
        }

        return tt;  // Return total frames sent and resent
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.print("Total number of frames: ");
        int tf = scanner.nextInt();

        System.out.print("Window Size: ");
        int N = scanner.nextInt();

        int i = 1;  // Initial frame index
        int tt = selectiveRepeatTransmission(i, N, tf);  // Run selective repeat transmission

        System.out.println("Total number of frames which were sent and resent: " + tt);
        scanner.close();
    }
}
